gax/src/main/java/com/google/api/gax/rpc/mtls/MtlsUtils.java

```diff
@@ -0,0 +1,178 @@
+ */
+
+package com.google.api.gax.rpc.mtls;
+
+import com.google.api.client.json.JsonParser;
+import com.google.api.client.json.jackson2.JacksonFactory;
+import com.google.api.client.util.SecurityUtils;
+import com.google.api.core.BetaApi;
+import com.google.common.annotations.VisibleForTesting;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.security.GeneralSecurityException;
+import java.security.KeyStore;
+import java.util.List;
+
+/** Utilities for mutual TLS. */
+@BetaApi
+public class MtlsUtils {
+  @VisibleForTesting
+  static class DefaultMtlsProvider implements MtlsProvider {
+    private static final String DEFAULT_CONTEXT_AWARE_METADATA_PATH =
+        System.getProperty("user.home") + "/.secureConnect/context_aware_metadata.json";
+
+    /** GOOGLE_API_USE_CLIENT_CERTIFICATE environment variable. */
+    public static final String GOOGLE_API_USE_CLIENT_CERTIFICATE =
+        "GOOGLE_API_USE_CLIENT_CERTIFICATE";
+
+    /** GOOGLE_API_USE_MTLS_ENDPOINT environment variable. */
+    public static final String GOOGLE_API_USE_MTLS_ENDPOINT = "GOOGLE_API_USE_MTLS_ENDPOINT";
+
+    interface EnvironmentProvider {
+      String getenv(String name);
+    }
+
+    static class SystemEnvironmentProvider implements EnvironmentProvider {
+      @Override
+      public String getenv(String name) {
+        return System.getenv(name);
+      }
+    }
+
+    DefaultMtlsProvider() {
+      this(new SystemEnvironmentProvider(), DEFAULT_CONTEXT_AWARE_METADATA_PATH);
+    }
+
+    private EnvironmentProvider envProvider;
+    private String metadataPath;
+
+    /** Cache the key store so we don't create a new one every time getKeyStore is called. */
+    private KeyStore keyStore = null;
+
+    @VisibleForTesting
+    DefaultMtlsProvider(EnvironmentProvider envProvider, String metadataPath) {
+      this.envProvider = envProvider;
+      this.metadataPath = metadataPath;
+    }
+
+    @Override
+    public boolean useMtlsClientCertificate() {
+      String useClientCertificate = envProvider.getenv(GOOGLE_API_USE_CLIENT_CERTIFICATE);
+      return "true".equals(useClientCertificate);
+    }
+
+    @Override
+    public String useMtlsEndpoint() {
+      return envProvider.getenv(GOOGLE_API_USE_MTLS_ENDPOINT);
+    }
+
+    @Override
+    public String getKeyStorePassword() {
+      return "";
+    }
+
+    @Override
+    public KeyStore getKeyStore() throws IOException, GeneralSecurityException {
+      if (keyStore != null) {
+        // return the cached key store
+        return keyStore;
+      }
+      try {
+        // Load the cert provider command from the json file.
+        InputStream stream = new FileInputStream(metadataPath);
+        List<String> command = extractCertificateProviderCommand(stream);
+
+        // Run the command and timeout after 1000 milliseconds.
+        Process process = new ProcessBuilder(command).start();
```

Comment:
Do we really need to start an external process to do what we need to do here? Looking at the tests it seems like it will start `cat` process, which is system dependent tool, and is not necessary since we can do the same thing in a platform-independent java native way.

Reply:
Yes, we need to run the command provided in `context_aware_metadata.json`. It is not a `cat` process, it is a binary installed on the system, for gLinux, it is `/opt/google/endpoint-verification/bin/apihelper`.